### ISSUE TITLE
Some related or extended questions

### ISSUE THREAD

**Issue by nviwch** [Jeff-sjtu/res-loglikelihood-regression#4]
I implemented a tensorflow keras version for 2D single point regression with some success. The mu is more accurate (sometimes better than ground truth due to human label error) and sigma do have a meaningful representation. 

But I have some questions, 

I would like to know if sigma bounded by (0, 1) is solely for the ease of classifying missing point or there is another reason. I tried using softplus in sigma so that it is > 0, the result is also very good and the sigma no longer squeezed between (0, 1), a blurry/noisy image can have very high sigma > 4.

As it is possible that the key point is missing in the image, I used a simple classification output previously, i.e. output one more sigmoid value to represent if the point exists. Now I have sigma, but few normal cases' sigma do overlap with missing point cases' sigma around ~0.5 (using softplus). Previously, I can feed missing point image to train, but now I can't? Just asking for some advice, how do I improve missing point detection?

**Comment by Jeff-sjtu** [comment]
Hi @nviwch,

For the first question, we use `sigmoid` just to make sure the value of `sigma` is positive, and the confidence score is positive. In our implementation, the coordinate is normalized by the image size, i.e. the coordinate within the image is between (-0.5, 0.5). Thus we think bounding `sigma` by (0, 1) works in most cases. Still, I think `softplus` is also a good alternative to `sigmoid`.

For the second question, I think you can still feed the missing point to train the model. Since we use a regression-based method, the network can predict the coordinate outside the input image. For example, the output coordinate can be larger than 0.5 or smaller than -0.5. In our supplementary files, we also show how our method works when facing truncation (i.e. so points are missing in the image). Our method has the ability to infer the missing points.

**Comment by nviwch** [comment]
I think my problem is different from truncation/occulsion because the points simply does not exist. Any value >0.5 or <-0.5 would be valid for missing point. 

For example, detecting centre of a circle, if part of the circle still presents in the image, we can still infer the centre position by the arc. But the circle simply does not exist in the image. How do I formulate it when all values outside 0.5 or -0.5 are valid/correct? This is different from a precise ground truth coordinate.

**Comment by Jeff-sjtu** [comment]
Oh, sorry for misunderstanding your question. In the test time, do you need to classify whether the point exist? If not, you can just multiply a zero-weight for the missing point during training. In our case, if some body keypoints are missing, i.e. not labelled by human annotators, we just multiply 0 to these points. If you need to classify, you can still use the 0 loss weight for regression loss and use an additional classification loss. I think `sigma` can provide some clues for whether the point exists, but it may be not very accurate.

**Comment by nviwch** [comment]
I tried to output two sigma instead of one, adding this flexibility, **lower the loss a bit more**. bar_mu = (hat_mu - gt) / sigma, the division would then be element-wise division, nf_loss = log(sigma) - log_phi would have one more dimension, the log Q still use the same laplace distribution but replaced with sigma having two dimensions (multivariate laplace distribution is too complicated that I didn't try).

I also added a classification head to my model. I just add the binary cross entropy to the original loss (as bce has the same scale/meaning as negative loglikelihood? What do you think). The result is quite good. 

But now I am thinking the contribution of sigma is like adding flexibility to the model by letting flow model to learn the distribution of (hat_mu, gt_mu, sigma), in this case it is the residue of the Q function and z-score distribution. But it does not have to be the z-score, so I am thinking how to use it in classifcation.

Classification can also be beneficial by your approach because human only label 0 or 1 and we don't know how to label between 0 to 1 _(may be the flow model can help us?)_ . I am now doing it binary classifcation, i.e. outputting one value from 0 to 1. The easiest way would be just do it like regression ((hat_mu - gt) / sigma). But usually we use binary crossentropy as the loss, it is strange to do (hat_score - gt) / sigma.

**Comment by Jeff-sjtu** [comment]
Hi @nviwch, I am glad to see RLE works in your project.

1. The cross-entropy loss is also derived from maximum likelihood estimation.
2. I think `sigma` cannot be directly used in classification. Since it describes the standard deviation, it is not calibrated for the probability that represents what are the chances of the point is visible/missing.
3. We are also interested in how to apply our approach in classification. There are some differences here. It is natural to apply re-parameterization in regression, but not in classification. In regression, the output can be treated as a continuous distribution, while there is a discrete distribution in classification. I do not have a really good solution right now, but I will keep working on that.

**Comment by nviwch** [comment]
期待分類網絡上用RLE
